### PR TITLE
[mellanox] Extend Mellanox FW utils with CPLD update

### PR DESCRIPTION
**Pull request by nazariig**
Signed-off-by: Nazarii Hnydyn <nazariig@mellanox.com>

Extend Mellanox FW utils with CPLD update feature

**- What I did**
* Added support for CPLD upgrade to Mellanox FW utility

**- How I did it**
* Updated Mellanox FW utility

**- How to verify it**
1. mlnx-fw-upgrade.sh --upgrade --cpld # Regular CPLD update flow
2. UPDATE_MLNX_CPLD_FW=1 mlnx-fw-upgrade.sh --upgrade # Force CPLD refresh only

**- Description for the changelog**
1. Improved error handling/resources cleanup/verbosity of Mellanox FW utility

**- A picture of a cute animal (not mandatory but encouraged)**
```
      .---.        .-----------
     /     \  __  /    ------
    / /     \(  )/    -----
   //////   ' \/ `   ---
  //// / // :    : ---
 // /   /  /`    '--
//          //..\\
       ====UU====UU====
           '//||\\`
             ''``
```

### CODE REVIEW

**Comment by sujinmkang** [comment]
please rebuild all!

**Comment by sujinmkang** [comment]
rebuild all please

**Comment by sujinmkang** [comment]
redo all please

**Comment by sujinmkang** [comment]
retest this please

File: platform/mellanox/mlnx-fw-upgrade.j2 [hunk 1]
```diff
@@ -47,43 +70,119 @@ function ParseArguments() {
...
+        sleep 2s
+        return "${EXIT_SUCCESS}"
+    fi
+
+    for i in "${SPIN[@]}"; do
+        echo -ne "\r[${i}] ${1}"
+        sleep 0.5s
+    done
+
+    echo -ne "\033[1K\r"
```
Comment (qiluo-msft) [inline platform/mellanox/mlnx-fw-upgrade.j2]: Add some code comment?

**Comment by sujinmkang** [comment]
retest this please

File: platform/mellanox/mlnx-fw-upgrade.j2 [hunk 2]
```diff
@@ -93,51 +192,148 @@ function UpgradeFW() {
...
+
+    RunCmd "${BURN_CMD} -i ${_FW_FILE}"
+}
+
+function UpgradeCPLDFW_Worker() {
+    local -r _CPLD_BURN_FILE="${1}"
+    local -r _CPLD_REFRESH_FILE="${2}"
+    local -r _ASIC_DEV="$(find /dev/mst -iname '*_pciconf0')"
+
+    if [[ "${UPDATE_MLNX_CPLD_FW}" = "1" ]]; then
```
Comment (sujinmkang) [inline platform/mellanox/mlnx-fw-upgrade.j2]: UPDATE_MLNX_CPLD_FW is not passed for sudo reboot during firmware update. can we have a file to indicate the cpld firmware is updated and refer to the file for the refresh command execution?

File: platform/mellanox/mlnx-fw-upgrade.j2 (hunk 1 above)
Comment (nazariig) [reply platform/mellanox/mlnx-fw-upgrade.j2]: @qiluo-msft done

File: platform/mellanox/mlnx-fw-upgrade.j2 (hunk 2 above)
Comment (nazariig) [reply platform/mellanox/mlnx-fw-upgrade.j2]: @sujinmkang answered via the mail

**Comment by sujinmkang** [comment]
retest mellanox please!

**Comment by sujinmkang** [comment]
Retest mellanox please.

**Comment by sujinmkang** [comment]
Retest Mellanox Please!

**Comment by sujinmkang** [comment]
retest mellanox please

**Pull request merged**